File: lib/domain/qr_code/attribute_resistance_reverse_calculator.dart
import '../master_data/master_data.dart';
import 'attribute_resistance.dart';
import 'attribute_resistance_calculator.dart';

/// Searches [masterData] for a single body color (with or without SP), a
/// pair of distinct colors, or a same-color pair whose calculated
/// attribute resistance matches this list exactly.
///
/// Returns the first matching `(bodyColors, isSpColor)` combination found,
/// or null if none of the solo, distinct-pair, or same-pair candidates
/// produce that result.
extension AttributeResistanceReverseLookup on List<AttributeResistance> {
  BodyColorSelection? findColorCombination(MasterData masterData) {
    final target = _toMap(this);
    final colorIds = masterData.bodyColorResistanceRules
        .map((rule) => rule.colorId)
        .toList();

    for (final colorId in colorIds) {
      for (final isSpColor in [false, true]) {
        final candidate = (bodyColors: [colorId], isSpColor: isSpColor);
        if (_matches(candidate, target, masterData)) {
          return candidate;
        }
      }
    }

    for (final colorId in colorIds) {
      final candidate = (bodyColors: [colorId, colorId], isSpColor: false);
      if (_matches(candidate, target, masterData)) {
        return candidate;
      }
    }

    for (var i = 0; i < colorIds.length; i++) {
      for (var j = i + 1; j < colorIds.length; j++) {
        final candidate = (
          bodyColors: [colorIds[i], colorIds[j]],
          isSpColor: false,
        );
        if (_matches(candidate, target, masterData)) {
          return candidate;
        }
      }
    }

    return null;
  }
}

bool _matches(
  BodyColorSelection candidate,
  Map<String, int> target,
  MasterData masterData,
) {
  final result = candidate.calculateAttributeResistance(masterData);
  return _mapEquals(_toMap(result), target);
}

Map<String, int> _toMap(List<AttributeResistance> list) => {
  for (final resistance in list) resistance.attributeId: resistance.value,
};

bool _mapEquals(Map<String, int> a, Map<String, int> b) {
  if (a.length != b.length) return false;
  for (final entry in a.entries) {
    if (b[entry.key] != entry.value) return false;
  }
  return true;
}

File: modules/server/src/domain/resistance/attribute-resistance-reverse-calculator.ts
import { calculateAttributeResistance } from './attribute-resistance-calculator';
import type { AttributeResistance, BodyColorSelection, ResistanceMasterData } from './types';

/**
 * Ports `AttributeResistanceReverseLookup.findColorCombination` from
 * `lib/domain/denpa_men/attribute_resistance_reverse_calculator.dart`.
 * Searches for a single body color (with or without SP), a same-color
 * pair, or a distinct-color pair whose calculated attribute resistance
 * matches `target` exactly. Returns the first match, or `null`.
 */
export function findColorCombination(
  target: AttributeResistance[],
  masterData: ResistanceMasterData,
): BodyColorSelection | null {
  const targetMap = toMap(target);
  const colorIds = masterData.bodyColorResistanceRules.map((rule) => rule.colorId);

  const candidates: BodyColorSelection[] = [
    ...colorIds.map((colorId): BodyColorSelection => ({ bodyColors: [colorId], isSpColor: false })),
    ...colorIds.map((colorId): BodyColorSelection => ({ bodyColors: [colorId], isSpColor: true })),
    ...colorIds.map((colorId): BodyColorSelection => ({ bodyColors: [colorId, colorId], isSpColor: false })),
  ];
  for (let i = 0; i < colorIds.length; i++) {
    for (let j = i + 1; j < colorIds.length; j++) {
      candidates.push({ bodyColors: [colorIds[i], colorIds[j]], isSpColor: false });
    }
  }

  for (const candidate of candidates) {
    const result = calculateAttributeResistance(candidate, masterData);
    if (mapEquals(toMap(result), targetMap)) {
      return candidate;
    }
  }

  return null;
}

function toMap(list: AttributeResistance[]): Map<string, number> {
  return new Map(list.map((resistance) => [resistance.attributeId, resistance.value]));
}

function mapEquals(a: Map<string, number>, b: Map<string, number>): boolean {
  if (a.size !== b.size) return false;
  for (const [key, value] of a) {
    if (b.get(key) !== value) return false;
  }
  return true;
}
